File: Difficulty: Medium/Maximum Bipartite Matching/maximum-bipartite-matching.py
class Solution:
    def maximumMatch(self, G):
        # Initialize the graph, number of people, and number of jobs
        self.graph = G
        self.num_people = len(G)
        self.num_jobs = len(G[0])
        
        # Call the function to get the maximum number of job matches
        return self.findMaxBPM()
        
    
    def findMaxBPM(self):
        # Initialize an array to track job assignments, -1 means no job assigned
        job_assigned = [-1] * self.num_jobs
        
        max_matches = 0
        # Try to assign a job for each person
        for person in range(self.num_people):
            # Keep track of jobs already visited in the current DFS attempt
            visited_jobs = [False] * self.num_jobs
            
            # If a job is assigned for the person, increment the result
            if self.tryToAssignJob(person, job_assigned, visited_jobs):
                max_matches += 1

        return max_matches
    
    def tryToAssignJob(self, person, job_assigned, visited_jobs):
        # Check all jobs for the current person
        for job in range(self.num_jobs):
            # If the person can do the job and it hasn't been visited yet
            if self.graph[person][job] and not visited_jobs[job]:
                visited_jobs[job] = True
                
                # If the job is not assigned or we can find an alternative match for the previously assigned person
                if job_assigned[job] == -1 or self.tryToAssignJob(job_assigned[job], job_assigned, visited_jobs):
                    job_assigned[job] = person
                    
                    return True
                    
        return False

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    T = int(input())
    for i in range(T):
        m, n = map(int, input().strip().split())
        G = []
        for i in range(m):
            G.append(list(map(int, input().strip().split())))
        obj = Solution()
        ans = obj.maximumMatch(G)
        print(ans)
        print("~")

# } Driver Code Ends

File: Difficulty: Medium/Maximum Bipartite Matching/maximum-bipartite-matching.js
//{ Driver Code Starts
//Initial Template for javascript

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(
        string => { return string.trim(); });

    main();
});

function readLine() {
    return inputString[currentLine++];
}

function printList(res, n) {
    let s = '';
    for (let i = 0; i < n; i++) {
        s += res[i];
        s += " ";
    }
    console.log(s);
}


function main() {
    let t = parseInt(readLine());
    let i = 0;
    for (; i < t; i++) {
        let input_ar1 = readLine().split(' ').map(x => parseInt(x));
        let a = input_ar1[0];
        let b = input_ar1[1];
        let arr = [];
        for (let i = 0; i < a; i++) {
            let s = readLine().split(' ');
            let row = new Array(b);
            for (let j = 0; j < b; j++) {
                row[j] = parseInt(s[j]);
            }
            arr.push(row);
        }
        let obj = new Solution();
        let res = obj.maximumMatch(arr, a, b);
        console.log(res);

    
console.log("~");
}

}
// } Driver Code Ends


//User function Template for javascript
/**
 * @param {number[][]} G
 * @returns {number}
*/
class Solution {
    /**
    * @param {number[][]} G
    * @returns {number}
    */
    maximumMatch(G) {
        const M = G.length;  // Number of people (rows in the graph)
        const N = G[0].length;  // Number of jobs (columns in the graph)
        return this.fun(G, M, N);
    }

    /**
    * @param {number[][]} G
    * @param {number} M
    * @param {number} N
    * @returns {number}
    */
    fun(G, M, N) {
        const match = Array(N).fill(-1);  // Stores job assignments (-1 means unassigned)
        let res = 0;  // Result to store the count of maximum matches

        for (let u = 0; u < M; u++) {  // Iterate for each person
            const seen = Array(N).fill(false);  // Mark all jobs as unvisited

            // If a job can be found for person u
            if (this.bpm(G, u, seen, match, M, N)) {
                res += 1;
            }
        }
        return res;
    }

    /**
    * @param {number[][]} G
    * @param {number} u
    * @param {boolean[]} seen
    * @param {number[]} match
    * @param {number} M
    * @param {number} N
    * @returns {boolean}
    */
    bpm(G, u, seen, match, M, N) {
        for (let v = 0; v < N; v++) {  // Try every job
            if (G[u][v] === 1 && !seen[v]) {  // If the job is available and not visited
                seen[v] = true;

                // If job v is not assigned or if the current applicant has no other job
                if (match[v] < 0 || this.bpm(G, match[v], seen, match, M, N)) {
                    match[v] = u;  // Assign job v to person u
                    return true;
                }
            }
        }
        return false;  // No job found for person u
    }
}
